### ISSUE TITLE
Question about design of CellValue and CellMatrix

### ISSUE THREAD

**Issue by dmenendez-gruposantander** [xlw/xlw#14]
Hi, long time xlw user here, happy to see the project in github.

My question is twofold:
1) why is CellValue a virtual base class? Is anyone out there actually using something different than MJCellValue?
2) why does CellMatrix use pimpl and why is CellMatrix_pimpl_abstract a virtual base class? Again, is anyone using something different that MJCellMatrix?

It seems to me the design would be much simpler if we just had everything concrete, no polymorphism and no pimpl. Also, everybody is paying for the possibility of having their own implementation, when I don't see much value to do that.

TLDR; my objective would be to add move semantics to CellMatrix and CellValue, and that would be much easier if those were plain value-based classes.

**Comment by keithalewis** [comment]
My sentiments exactly. The SDK XLOPER/XLOPER12 structs already represent both cells and ranges. 
I hope the xlw group doesn't mind some friendly A/B comparison, but it seems a better approach is to publicly inherit from these and add C++ members function to make them into proper value types.
Here's my approach to that: [oper.h](https://github.com/xlladdins/xll/blob/master/xll/oper.h).

**Comment by dmenendez-gruposantander** [comment]
> My sentiments exactly. The SDK XLOPER/XLOPER12 structs already represent both cells and ranges.
> I hope the xlw group doesn't mind some friendly A/B comparison, but it seems a better approach is to publicly inherit from these and add C++ members function to make them into proper value types.
> Here's my approach to that: [oper.h](https://github.com/xlladdins/xll/blob/master/xll/oper.h).

Well, I was thinking just to improve these implementation details, but in essence leave the public interface of CellValue and CellMatrix unchanged, just make them non-virtual, plain value-based classes being able to represent *owning* cells and ranges independent of Excel, and not backed by Excel data structures.

**Comment by keithalewis** [comment]
Are you using the xlw library to write Excel add-ins, or are you using it as a general purpose library for variants and 2-dimensional array types that happen to be restricted to the same types that Excel uses?

**Comment by dmenendez-gruposantander** [comment]
I had Excel add-ins using xlw, that evolved over time to get an independent life as standalone libraries that now happen to use Excel-like structures in some of their interfaces. So yes  ^_^

In my view, xlw could separate two sub-libraries: the data structures (CellValue and CellMatrix) and the conversion from/to XLOPER structs.

**Comment by Ninds** [comment]
Hi,
Just picking this up after quite a long time away,
There are quite a few consumers that need to build their binaries against CellMatrix & CellValue that are not backed by Excel types. Consider for example wanting to build a version of your xll that can be used as part of a regression batch.
